### ISSUE TITLE
uncertain regions Loss  parameter sigma

### ISSUE THREAD

**Issue by jason-xiao-yi** [luckybird1994/HQSOD#1]
Hello, author. May I ask how the sigma parameter in the loss of uncertain region in the paper was obtained?  Is it to be learned as a learnable parameter, or is it to be learned as a distribution factor of a deterministic region?  In the final loss function, adding logsigma to each pixel would cause the second term to be too large

**Comment by luckybird19941994** [comment]
> Hello, author. May I ask how the sigma parameter in the loss of uncertain region in the paper was obtained? Is it to be learned as a learnable parameter, or is it to be learned as a distribution factor of a deterministic region? In the final loss function, adding logsigma to each pixel would cause the second term to be too large

give me your e-mail, I send the code to you.

**Comment by chiyuan-IVIPLab** [comment]
Dear author. Compared with other SOD tasks, the method of uncertainty loss made a great improvement on the blur edge process(especially on Bμ and BDE) may I see some details about the uncertainty loss? My e-mail is: **1140395492@qq.com**, thank you so much.

**Comment by jason-xiao-yi** [comment]
作者实际采用额不确定损失函数已跟原论文不同，原论文函数存在极大漏洞，使用这个函数对模型并没有实际改善，但不确定性建模仍然值得研究
以下是我跟作者来往的邮件记录及作者发给我的损失代码，包含了对不确定性函数的想法及探讨。虽然作者的函数确实存在问题，不过作者大大人还是挺值得钦佩，会直面人们的疑问。
[image: image.png]
import torch.nn.functional as F
import numpy as np
import torch
import copy
import random

def calculate(val, mean, var):
    '''
    input: pre, mask, sigma
    out:公式: mean(|pre- mask|**2 / (2*sigma) + 0.5*根号(2pi*sigma))
    '''

    temp_var = torch.log(var)
    temp_var = torch.exp(temp_var)   # 此步已无意义，猜测作者最后根据结果，让函数返回去拟合模型结果
    difference = (val-mean) ** 2.0

    loss1 = 0.5 * difference / temp_var
    loss2 = 0.5 * torch.sqrt(2.0 * np.pi * var)
    loss = loss1 + loss2
    return loss.mean()

def _uncertain_loss(mask,prediction,uncertain_para,trimap):

    N,C,W,H = mask.shape

    #其实就是开始阶段，不怎么好训练，容易出现极大或者极小值
    uncertain_para = torch.sigmoid( uncertain_para ).clamp(0.01,0.99) #
限幅。clamp将input的值限制在[min, max]之间，并返回结果。out (Tensor, optional) – 输出张量，一般用不到该参数

    #修改预测，不确定参数，mask 维度shape,确保一致
    prediction = prediction.view(N,W*H)
    uncertain_para = uncertain_para.view(N,W*H)
    mask = mask.view(N,W*H)
    #获取trimap并修改shape?en? 作者sod阶段输出三通道，则unknown区域信息应该保存在第2通道
    index = trimap[:, 1:2, :, :].float()
    index = index.view(N, W * H)

    loss_uncertain = torch.tensor(0.0)  # 初始化uncertain loss
    uncertrain_cords = torch.eq(index,1)  #
对两个张量Tensor进行逐元素的比较，若相同位置的两个元素相同，则返回True；若不同，返回False。
这一步是获得unknown区域位置，矩阵元素为：unknown区域为true,其余为true
    uncertain_prediction = prediction[uncertrain_cords]  #
获得unknown区域的预测。维度变为一维向量（但是维度会改变吧）   还能这样写？？？经测试，两个矩阵shape相同时，返回索引位置为true的元素
    if uncertain_prediction.shape[0] != 0:
        #获得mask,pre 对应unknown区域的信息，shape:一维数组
        uncertrain_conf = uncertain_para[uncertrain_cords]
        uncertrain_mask = mask[uncertrain_cords]
        loss_uncertain = calculate(uncertain_prediction,uncertrain_mask,
uncertrain_conf)

    index_white = trimap[:,2:3,:,:].float()
    index_white = index_white.view(N,W*H)
    loss_white_uncertain = torch.tensor(0.0)

    uncertain_white_cords = torch.eq(index_white,1)
    uncertain_white_prediction = prediction[uncertain_white_cords]
    if uncertain_white_prediction.shape[0] != 0:

        uncertrain_white_conf = uncertain_para[uncertain_white_cords]
        uncertrain_white_mask = mask[uncertain_white_cords]

        white_num = uncertain_white_prediction.shape[0]
        id_list_white = np.random.permutation( white_num )
        id_list_white = list(id_list_white)
        sample_num = min(white_num,2000)
        slice = random.sample(id_list_white,sample_num)
        uncertrain_white_conf = uncertrain_white_conf[slice]
        uncertrain_white_mask = uncertrain_white_mask[slice]
        uncertain_white_prediction = uncertain_white_prediction[slice]
        loss_white_uncertain = calculate(uncertain_white_prediction,
uncertrain_white_mask,uncertrain_white_conf)

    index_black = trimap[:, 0:1, :, :].float()
    index_black = index_black.view(N,W*H)
    loss_black_uncertain = torch.tensor(0.0)

    uncertain_black_cords = torch.eq(index_black, 1)
    uncertain_black_prediction = prediction[uncertain_black_cords]
    if uncertain_black_prediction.shape[0] != 0:

        uncertrain_black_conf = uncertain_para[uncertain_black_cords]
        uncertrain_black_mask = mask[uncertain_black_cords]

        black_num = uncertain_black_prediction.shape[0]
        id_list_black = np.random.permutation(black_num)
        id_list_black = list(id_list_black)
        sample_num = min(black_num,1000)
        slice = random.sample(id_list_black,sample_num)
        uncertrain_black_conf = uncertrain_black_conf[slice]
        uncertrain_black_mask = uncertrain_black_mask[slice]
        uncertain_black_prediction = uncertain_black_prediction[slice]
        loss_black_uncertain = calculate(uncertain_black_prediction,
uncertrain_black_mask,uncertrain_black_conf)

    loss = loss_uncertain + loss_black_uncertain + loss_white_uncertain
    return loss

if __name__=='__main__':

    trimap = torch.randn( (2,2) ).to(torch.long)
    trimap[0,0] = 1
    trimap[0,1] = 0
    trimap[1,0] = 2
    trimap[1,1] = 2
    trimap = F.one_hot(trimap,num_classes=3).permute(2,0,1).float()
    trimap = trimap.unsqueeze(0)

    index = trimap[:,0:1,:,:]




    mask = torch.randn( (1,1,2,2))
    prediction = torch.randn((1,1,2,2))

    uncertain_para = torch.randn((1,1,2,2))
    device = torch.device("cuda:1" if torch.cuda.is_available() else "cpu")

    _uncertain_loss(mask,prediction,uncertain_para,trimap)






jokerhcy ***@***.***> 于2022年1月6日周四 14:36写道：

**Comment by poorpaper** [comment]
Dear author. I have read the article and found the uncertainty loss really made a great improvement on the boundary of some objects. But I didn't understand how can I get the sigma in that loss. Could you please send me the code? E-mail: 1047348834@qq.com . Thank you.

**Comment by shkarupa-alex** [comment]
Could you please send me the code of uncertainty loss? E-mail: [shkarupa.alex@gmail.com](mailto:shkarupa.alex@gmail.com) . Thanks.

**Comment by xiechenxi99** [comment]
Could you please send me the code of uncertainty loss? E-mail: [1072593143@qq.com](mailto:1072593143@qq.com) . Thanks.

**Comment by EricWu-ZL** [comment]
Could you please send me the code of uncertainty loss? E-mail: [eric3772443@gmail.com](mailto:eric3772443@gmail.com) . Thanks.
